Workspace flow + TestSpec generation

tjfulle opened the pull request:
This PR supersedes #66 and includes all its details.  Additionally, it includes a major change to the generation of test cases as described in [spec-case.md](https://github.com/sandialabs/canary/blob/tjfulle/specs/docs/design/spec-case.md)

tjfulle committed:
save each test generator

tjfulle committed:
quick commit

tjfulle committed:
another update

tjfulle committed:
no more canary exec, use canary run

tjfulle committed:
rempove broke import

tjfulle committed:
use stage instead of lock

mdmosby committed:
add fairly comprehensive tutorial document for the prototype workflow

tjfulle committed:
more work with tagging/staging adding gc

tjfulle committed:
create markdown reports

tjfulle committed:
html and json reports

tjfulle committed:
finish up some reporting

tjfulle committed:
removing usage of session in favor of repo

tjfulle committed:
change name repo->workspace.  make tag objects lightweight.  log and location should now work

tjfulle committed:
Update gc.py

mdmosby committed:
draft refined workspace command spec

tjfulle committed:
some clean up to workspace code (better error messsages and predicate checking).  begin to look for `start` argument of pathspec

tjfulle committed:
fixing status and hpc run/exec

tjfulle committed:
Can rerun failed!

tjfulle committed:
dont tag unless explicitly asked to do so.

tjfulle committed:
add tag -d to remove tags and fix some bugs

tjfulle committed:
config:view option lets one change the name of the view, or not generate it.  fixed up garbage collecting

tjfulle committed:
Better garbage collecting

tjfulle committed:
tests pass!

tjfulle committed:
cmake tests pass

tjfulle committed:
fix canary hpc tests

tjfulle committed:
combine hpc and batch commands - batch is an alias for hpc

tjfulle committed:
fix show excluded tests

tjfulle committed:
Update workspace.py

tjfulle committed:
Update select.py

tjfulle committed:
Update workspace.py

tjfulle committed:
first pass at test specs

tjfulle committed:
spec commons

tjfulle committed:
DraftSpec -> ResolvedSpec -> TestSpec

tjfulle committed:
adding new process pool executor

tjfulle committed:
can run pyt files with new process pool

tjfulle committed:
more work on test specs

tjfulle committed:
another update

tjfulle committed:
fix codes put on tests

tjfulle committed:
some more linting

tjfulle committed:
another update

tjfulle committed:
more type checking

tjfulle committed:
almost cmake

tjfulle committed:
cmake works and running from file in test session works

tjfulle committed:
run batches

tjfulle committed:
fix instantiation of execpace

tjfulle committed:
run in a current session

tjfulle committed:
fix some typing error

tjfulle committed:
reenable rebaseline

tjfulle committed:
more robust pool processing

tjfulle committed:
fixing some more subcommands

tjfulle committed:
get ty check to work on src.  still not on tests

tjfulle committed:
one more

tjfulle committed:
resourcequeueexecutor

tjfulle committed:
ruff

tjfulle committed:
remove some config methods

tjfulle committed:
config is a simple dict

tjfulle committed:
add scratch section

tjfulle committed:
better config funciton name: update -> write-new

tjfulle committed:
ruff

tjfulle committed:
Merge branch 'main' into tjfulle/specs

tjfulle committed:
simplify config to not have config: key

tjfulle committed:
toward getting tests running

tjfulle committed:
some more work on signals

tjfulle committed:
JobProtocol

tjfulle committed:
plugin batching to newer signal processing

tjfulle committed:
some cleanup

tjfulle committed:
adding protocols file and refactor resource queue

tjfulle committed:
Update queue.py

tjfulle committed:
clean up queue a bit

tjfulle committed:
move batching to conductor

tjfulle committed:
add batchjob module

tjfulle committed:
stub

tjfulle committed:
one more

tjfulle committed:
one more

tjfulle committed:
Adding batch executor

tjfulle committed:
Update batchexec.py

tjfulle committed:
Update batchexec.py

tjfulle committed:
Update json_helper.py

tjfulle committed:
Update config.py

tjfulle committed:
Update queue.py

tjfulle committed:
High level subprocess wrapper to print exc

tjfulle committed:
better signal handling in batch runner

tjfulle committed:
Adding command attribute when case is run

tjfulle committed:
add plugin hooks earlier in the call as part of config setup.  batch runner is a plugin call

tjfulle committed:
cleanup

tjfulle committed:
fix how resources are checked out and add a set_status method to testbatch

tjfulle committed:
core tests fixed!

tjfulle committed:
fix vvtest unit tests

mdmosby committed:
`log`: fix FileNotFoundError logic

tjfulle committed:
cmake tests pass

mdmosby committed:
`describe`: get display_name from case (callable), not spec (property)

tjfulle committed:
fixed canary_hpc tests

tjfulle committed:
fix tar/compress for pushing to cdash

tjfulle committed:
make refresh robust

mdmosby committed:
do not execute tests that have status other than "READY"

mdmosby committed:
capture/log child process exceptions and mark the status as ERROR

mdmosby committed:
improve error reporting for `canary init` when the workspace already exists

mdmosby committed:
more robustly remove the workspace during setup

workspace.setup() calls remove then create. I was getting filesystem errors on `remove` I think due
to buffering/write speed of the lockfile. This change resolves those sporadic errors

tjfulle committed:
add testspec_modify and cdash artifacts bundled as one

tjfulle committed:
give with_traceback a concrete function interface

tjfulle committed:
fix bug in with_traceback that didnt pass kwargs through

tjfulle committed:
adding in create_scope config method for backward compatibility

tjfulle committed:
adding some more logging

tjfulle committed:
adding yet more logging

tjfulle committed:
adding yet more logging

tjfulle committed:
fix collectreport to only take specs

tjfulle committed:
add keywords to testcase

tjfulle committed:
speed up testspec resolve a little

tjfulle committed:
fixes:

- pyt/vvt: use subprocess not runpy.  runpy is still there but
  there is difficulty in properly redirecting output streams that
  needs to be resolved.
- fast/parallel dependency resolver

tjfulle committed:
fixing warnings found by bandit

tjfulle committed:
ruff format/check

tjfulle committed:
Limiting default number of workers

tjfulle committed:
cleaning up some logging

mdmosby commented on src/canary_cmake/__init__.py:
```diff
@@ -122,3 +163,20 @@ def canary_resource_pool_fill(config: canary.Config, pool: dict[str, dict[str, A
+
+@canary.hookimpl(wrapper=True)
+def canary_cdash_artifacts(case: canary.TestCase) -> Generator[None, None, list[dict[str, str]]]:
+    """Default implementation: return the test case's keywords"""
```
```suggestion
    """Default implementation: return the test case's artifact specifications"""
```
Should `canary.TestCase` use this Schema to validate the artifacts specification?

mdmosby commented on src/canary_cmake/__init__.py:
```diff
@@ -122,3 +163,20 @@ def canary_resource_pool_fill(config: canary.Config, pool: dict[str, dict[str, A
+    artifacts.extend([_ for _ in result if _])
+    for i, artifact in enumerate(artifacts):
+        artifacts[i] = schema.validate(artifact)
+    return artifacts
```
What is the expected behavior if `artifacts` include elements with duplicate `file` values and conflicting `when` values?

tjfulle committed:
fix incorrect log message

mdmosby reviewed (approved):
Minor things that can wait

tjfulle committed:
clear out queue if cancelled

tjfulle committed:
re-raise keyboard interrupt

tjfulle committed:
fixing vvtest tests

tjfulle committed:
ruff docs

mdmosby committed:
fix traceback when collecting generators from version-control paths

mdmosby committed:
fix implementation of ExecutionSpace.copy

`Path.hardlink_to` requires that `src` and `dst` are on the same
**filesystem** which is a more-stringent requirement than we've previously
had. Furthermore, `hardlink_to` does not follow copy symantics in that
modifying the hardlinked file _also modifies the source_! Which may be
unexpected behavior when one specifies `copy`.

tjfulle committed:
toward faster testcase discovery

tjfulle committed:
Parallelizing test discovery:

1. Generator plugins must now implement canary_collect_file_patterns that returns a list of glob patterns
   that the generator recognizes
2. During discovery, we walk the tree and collect all files matching patterns from above
3. After collection, we call another new plugin canary_collect_filter_files.  This plugin can filter
   out files from the list collected.  This is useful for the ctest plugin to remove all but the
   first CTestTestfile in a tree
4. After filtering, create test case generators in parallel.

Additionally, store all generator meta data in a single file.

With the above changes, time to collect generators for a large test suite went from 25-30 seconds to less
than 1 second.

tjfulle committed:
remove errant print statement

tjfulle committed:
modularize some logging prefix

mdmosby committed:
correct docstring

tjfulle committed:
use sql database to hold latest results

tjfulle committed:
move logging into workspace

mdmosby committed:
improve performance of git-based generator collection

mdmosby committed:
fix implementation of repo_ls and apply performance improvements

tjfulle committed:
use sqlite db for everything in workspace

mdmosby committed:
perf: find_duplicates with O(N) algorithm

from O(N**2)

tjfulle committed:
dont cache any workspace objects outside of database

tjfulle committed:
fix typehint

tjfulle committed:
fix up some logging

tjfulle committed:
fix some linting and ci errors

tjfulle committed:
fix canary-hpc tests

mdmosby committed:
copy file with metadata/permissions

mdmosby committed:
fix typo in progress message

mdmosby committed:
fix generator file collection from version control

mdmosby committed:
factor out method to get the filesystem root when passed a VC pathspec

mdmosby committed:
format

mdmosby committed:
split _get_specs_by_id into chunks to get arount query variable limit

mdmosby committed:
fix quoting

mdmosby committed:
remove limit on number of partial ids to match

mdmosby committed:
fix infinite loop

mdmosby committed:
handle time formatting in caching

mdmosby committed:
minor improvement to `canary info -t tag` output

mdmosby committed:
fix indexing bug -- raise correct exception

tjfulle committed:
several fixes to get large test suite running on slurm:

- properly retrieve dependent specs to help build up test case
- add execpath property to spec that can be updated
- add canary_select_modifyitems plugin hook
- more database query options

tjfulle committed:
Update run.py

tjfulle committed:
introduce collector

tjfulle committed:
consolidate all collection logic/code to _canary/collect.py

tjfulle committed:
add new collect/select/rules modules

tjfulle committed:
fix some tests

tjfulle committed:
some more reorg and flattening

tjfulle committed:
adding build module

tjfulle committed:
fix typing error

tjfulle committed:
adding some docs

tjfulle committed:
trying to pass tests

tjfulle committed:
fix mypy errors

tjfulle committed:
fix core tests

tjfulle committed:
fix tests

tjfulle committed:
ruff format

tjfulle committed:
export more objects in canary/__init__.py

tjfulle committed:
print output of broken canary_hpc test

tjfulle committed:
add collector.run method to match style of build/select

tjfulle committed:
remove testcase_modify.  use build_modifyitems(builder) instead.

mdmosby committed:
fix collection from version control

mdmosby committed:
improve error reporting when a TestCase cannot be found

tjfulle committed:
create canary_runtests wrapper function and Runner object

tjfulle committed:
canary_runtests impl should return a non-None

tjfulle committed:
fix several mypy errors

tjfulle committed:
cleaner handling of banner/no-banner

tjfulle committed:
Update spec-case.md

tjfulle committed:
adding runtime filtering

tjfulle committed:
removing session

tjfulle committed:
col fields in results db

tjfulle committed:
remove TestSpec object and use ResolvedSpec in its place

tjfulle committed:
replace collector.add_file_patterns with collector.add_generator

tjfulle committed:
replace execution_policy -> launcher

tjfulle committed:
maybe fix flux tests...

tjfulle committed:
save initial config mods to workspace config.  closes: #70

tjfulle committed:
fix liniting

tjfulle committed:
add rich status table to run

tjfulle committed:
limit number of rows to print during run

mdmosby committed:
rules: ignore `default_reason` paramater in reconstruction and implement `__str__`

mdmosby committed:
canary info: fix filtering and re-enable reporting of selection metadata

tjfulle committed:
adding rerun command and refactoring status

• canary run will always run a tag/spec from a new state
• canary rerun will run a tag/spec from the previous state.  the distinction
is really only meaningful when dealing with dependencies.  in a rerun scenario
options can be passed specifying that dependents should not be rerun, or downstreams
should not be rerun, etc.

tjfulle committed:
lint fixes

tjfulle committed:
fix unit tests

tjfulle committed:
fix unit test

tjfulle committed:
Update status.py

tjfulle committed:
fix history command

tjfulle committed:
fix some comments

tjfulle committed:
use rich to render info

tjfulle committed:
remove progress-bar stuff

tjfulle committed:
add more rich support and clean up canary run

tjfulle committed:
fix bug in status printing

tjfulle committed:
fix bug that failed to allow rerunning cases in place

tjfulle committed:
make implicit paramters a first class attribute of spec

tjfulle committed:
Update timeout.py

tjfulle committed:
remove extra logging when selector was empty

tjfulle committed:
improve selector logging

tjfulle committed:
fix two unit tests and changed implicit_parameters -> meta_parameters

mdmosby committed:
perf: `canary info` returns in under 1s

implement general DB table count method for when only counts and not data is needed

mdmosby committed:
Revert "perf: `canary info` returns in under 1s"

This reverts commit de62d485bc8a3deb10329aa74b85fc1c967017bf.

mdmosby committed:
teach WorkspaceDatabase to count rows in a table

perf: `canary info` returns in under 1s for SIERRA use case

When a workspace contains many thousands of tests, reconstructing them
from the database just to get their count is slow. A workspace with the
SIERRA test suite added took ~8s to return `canary info`.

mdmosby committed:
format tests (./bin/pre-commit)

mdmosby committed:
teach `canary describe` to describe a ResolvedSpec if a `TestCase` is not found

mdmosby committed:
correct meta_parameter handling

tjfulle committed:
adding nosec to formatted sql query

tjfulle committed:
remove spack colorize module from third_party.colorize and use rich

rationale: rich is used for table and live rendering and is a mature/well supported library

tjfulle committed:
fix bad escape sequences in string literal

tjfulle committed:
make case.refresh more robust

tjfulle committed:
adding more cols to results table so that canary status does not have to construct cases

tjfulle committed:
Adding canary generate command

generation and selection are fundamentally different things.  this formalizes
this by adding a canary generate command

mdmosby committed:
enable selections that do not mask anything out with the current generators

mdmosby committed:
export `Asset` type and move towards consistent interface of `Un/ResolvedSpec`

`UnresolvedSpec` now uses `assets` if it exists and falls back to `file_resources`.
This makes it possible to construct a `ResolvedSpec` and `UnresolvedSpec` with the same arguments.

mdmosby committed:
apply pre-commit format/typechecking

tjfulle committed:
fix mask propagation and add --only=changed

tjfulle committed:
cleanup sql queries to be more idiomatic
